### PR TITLE
[clang-format] Don't break between string literal operands of <<

### LINKED ISSUE #44363: A newline is added after a string token if the next 2 tokens are lessless and string
|  |  |
| --- | --- |
| Bugzilla Link | [45018](https://llvm.org/bz45018) |
| Version | trunk |
| OS | Windows NT |
| Attachments | [File to verify two streamed strings aren't split](https://user-images.githubusercontent.com/92601663/143760409-3d77f355-57f2-49dd-be1d-0dee0cdad6e7.gz) |

## Extended Description 

Given the code:

 QStringList() << "something" << "something" << variable;

If I then run clang-format with the default config file, the output is the following code:

 QStringList() << "something"
               << "something" << variable

What I expected, was that it stays on a single line, unless it exceeds the column limit and possible penalties. In this case, untouched.

It stays untouched when I add a variable in between:

 QStringList() << "something" << a << "something" << variable;

Looking at the code, I noticed that somebody intentionally introduced this to split endline characters in commit df28f7b8dd6a032515109de4ff5b4067be95da8e (currently at lib/Format/TokenAnnotator.cpp, line 3073).

In the attachment I added a file with a slightly different example that I used when debugging clang-format.

### PR DESCRIPTION
**Pull request by owenca**
Fixes #44363.

### CODE REVIEW

File: clang/lib/Format/TokenAnnotator.cpp [hunk 1]
```diff
@@ -5118,10 +5118,6 @@ bool TokenAnnotator::mustBreakBefore(const AnnotatedLine &Line,
     return true;
   if (Left.IsUnterminatedLiteral)
     return true;
-  if (Right.is(tok::lessless) && Right.Next && Left.is(tok::string_literal) &&
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/TokenAnnotator.cpp:5121]: But why was that added? Alone from Github I can't navigate to the commit which introduced it.

**Review by HazardyKnusperkeks — APPROVED** [review]

File: clang/lib/Format/TokenAnnotator.cpp (hunk 1 above)
Comment (rymiel) [reply clang/lib/Format/TokenAnnotator.cpp:5121]: https://github.com/llvm/llvm-project/commit/2603ee0dc6003be0fbe43038b34580864d1e0c85
Here, 11 years ago

Comment (HazardyKnusperkeks) [reply clang/lib/Format/TokenAnnotator.cpp:5121]: Thanks.

And the test doesn't need to be changed?

Comment (owenca) [reply clang/lib/Format/TokenAnnotator.cpp:5121]: No, it still passes.

**Comment by s1Sharp** [comment]
this is good to be respectful. Please, lets check my mr https://github.com/llvm/llvm-project/pull/69859 also

**Comment by s1Sharp** [comment]
I think adding the possibility of breaking '<<' is a good idea if it's optional, what do u think?

**Comment by owenca** [comment]
> I think adding the possibility of breaking '<<' is a good idea if it's optional, what do u think?

+1, but see also https://github.com/llvm/llvm-project/pull/69859#issuecomment-1776064714.

**Comment by petrhosek** [comment]
We have noticed that after this change, `clang-format` would behave differently depending on the input formatting.

(I'm using `--style="{BasedOnStyle: google, ColumnLimit: 0}"` in the examples below.)

For example, if the input is
```
return os << "ClientEnd<"
          << "test_anonymous::SomeProtocol"
          << ">(" << value.channel().get() << ")";
```
the output format would be the same before and after this change.

However, if the input is
```
return os << "ClientEnd<" << "test_anonymous::SomeProtocol" << ">(" << value.channel().get() << ")";
```
prior to this change we would reformat it as
```
return os << "ClientEnd<"
          << "test_anonymous::SomeProtocol"
          << ">(" << value.channel().get() << ")";
```
but after this change we would format it as
```
return os << "ClientEnd<" << "test_anonymous::SomeProtocol" << ">(" << value.channel().get() << ")";
```
that is the same as input.

Is that behavior intentional?

**Comment by owenca** [comment]
IMO, the new behavior is consistent with the [documentation](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#columnlimit) for `ColumnLimt: 0`.

**Comment by mydeveloperday** [comment]
I wish we'd made this removal an option rather than just removing it, what we are seeing is reasonably formatted json or xml being streamed out is now poorly written

```c++
osjson << "{\n"
       <<"  \"name\": \"value\",\n"
       <<"  \"key\": \"abc\", \n"
       <<" }";
```

now becomes

```c++
osjson << "{\n" <<"  \"name\": \"value\",\n <<"  \"key\": \"abc\",\n  << "}";
```

While i understand its correct for column width its not as nicer.

These rules about breaking between << were part of the original author rules for now to best try and format c++ and I sort of feel we should at least honour it via an option.

I also feel that this will create a large amount of complaints about us changing defaults.

**Comment by mydeveloperday** [comment]
See #86893 88483

**Comment by mydeveloperday** [comment]
see #88483

**Comment by owenca** [comment]
> I wish we'd made this removal an option rather than just removing it, what we are seeing is reasonably formatted json or xml being streamed out is now poorly written
> 
> ```c++
> osjson << "{\n"
>        <<"  \"name\": \"value\",\n"
>        <<"  \"key\": \"abc\", \n"
>        <<" }";
> ```
> 
> now becomes
> 
> ```c++
> osjson << "{\n" <<"  \"name\": \"value\",\n <<"  \"key\": \"abc\",\n  << "}";
> ```

What version was used? I've just tried clang-format 18.1.3, and it's totally fine:
```
$ cat lessless.cpp
osjson << "{\n"
       << "  \"name\": \"value\",\n"
       << "  \"key\": \"abc\", \n"
       << " }";
$ clang-format -version
clang-format version 18.1.3
$ clang-format lessless.cpp | diff lessless.cpp -
$ 
```

**Pull request merged**